src/main/java/io/nats/streaming/StreamingConnectionImpl.java

```diff
@@ -119,22 +123,32 @@ StreamingConnectionImpl connect() throws IOException, InterruptedException {
             this.hbSubject = this.newInbox();
             this.ackSubject = String.format("%s.%s", NatsStreaming.DEFAULT_ACK_PREFIX, this.nuid.next());
 
-            this.dispatcher = nc.createDispatcher(msg -> {
+            this.ackDispatcher = nc.createDispatcher(msg -> {
                 String subject = msg.getSubject();
+                if (this.ackSubject.equals(subject)) {
+                    this.processAck(msg);
+                }
+            });
 
+            this.heartbeatDispatcher = nc.createDispatcher(msg -> {
+                String subject = msg.getSubject();
                 if(this.hbSubject.equals(subject)) {
                     this.processHeartBeat(msg);
-                } else if (this.ackSubject.equals(subject)) {
-                    this.processAck(msg);
-                } else {
-                    this.processMsg(msg);
                 }
             });
 
-            this.dispatcher.subscribe(this.hbSubject);
-            this.dispatcher.subscribe(this.ackSubject);
+            this.messageDispatcher = nc.createDispatcher(msg -> {
+                this.processMsg(msg);
+            });
+
+            this.heartbeatDispatcher.subscribe(this.hbSubject);
+            this.ackDispatcher.subscribe(this.ackSubject);
+
+            this.heartbeatDispatcher.setPendingLimits(-1, -1);
```

Comment:
Not part of this PR, but maybe someday this would need to reference the NATS options.

Reply:
i am not sure what you mean?

Reply:
Should pending limits always be infinite?

Reply:
oh, agreed, we may want to surface these in the future. I think for ack and hb we may want to keep hidden. But for message queues there is definitely value in making visible - with the caveat that these will interact with other streaming settings. So, may not be something we ever get to if we can get jetstream going.